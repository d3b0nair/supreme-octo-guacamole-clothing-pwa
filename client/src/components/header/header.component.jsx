import React, { useEffect, useState } from "react";

import { connect } from "react-redux";
import { ReactComponent as Logo } from "../../assets/logo.svg";
import { createStructuredSelector } from "reselect";
import { selectCartHidden } from "../../redux/cart/cart.selector";
import { selectCurrentUser } from "../../redux/user/user.selector";
import { signOutStart } from "../../redux/user/user.actions";

import CartIcon from "../cart-icon/cart-icon.component";
import ScrollUpButton from "../scroll-up-button/scroll-up-button.component";
import CartDropdown from "../cart-dropdown/cart-dropdown.component";

import {
  HeaderContainer,
  LogoContainer,
  OptionsContainer,
  OptionLink,
  CartDropdownContainer,
} from "./header.styles";

export function Header({ currentUser, hidden, signOutStart }) {
  const [scrollDown, setScrollDown] = useState(false);
  const listenScrollEvent = () => {
    window.scrollY > 500 ? setScrollDown(true) : setScrollDown(false);
  };
  useEffect(() => {
    window.addEventListener("scroll", listenScrollEvent);
  });
  return (
    <>
      <HeaderContainer>
        <LogoContainer to="/">
          <Logo alt="logo" className="logo" />
        </LogoContainer>
        <OptionsContainer>
          <OptionLink to="/shop">SHOP</OptionLink>
          <OptionLink to="/shop">CONTACT</OptionLink>
          {currentUser ? (
            <OptionLink
              data-testid="signout-link"
              as="div"
              onClick={signOutStart}
            >
              SIGN OUT
            </OptionLink>
          ) : (
            <OptionLink to="/signin">SIGN IN</OptionLink>
          )}
          <CartIcon />
        </OptionsContainer>
        {hidden ? null : (
          <CartDropdownContainer data-testid="CartDropdown">
            <CartDropdown />
          </CartDropdownContainer>
        )}
      </HeaderContainer>
      {scrollDown ? <ScrollUpButton /> : null}
    </>
  );
}
const mapStatetoProps = createStructuredSelector({
  currentUser: selectCurrentUser,
  hidden: selectCartHidden,
});

const mapDispatchToProps = (dispatch) => ({
  signOutStart: () => dispatch(signOutStart()),
});

export default connect(mapStatetoProps, mapDispatchToProps)(Header);
