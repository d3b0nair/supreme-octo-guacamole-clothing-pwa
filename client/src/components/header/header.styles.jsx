import styled from "styled-components";
import { Link } from "react-router-dom";
import { global_padding_left_right } from "../../global.styles";

export const HeaderContainer = styled.header`
  height: 70px;
  width: 100%;
  display: flex;
  justify-content: space-between;
  margin-bottom: 25px;
  @media screen and (max-width: 800px) {
    height: 50px;
    padding: 10px;
    justify-content: space-around;
  }
  @media screen and (max-width: 300px) {
    height: 50px;
    padding: 5px;
    flex-wrap: wrap;
    flex-direction: row;
    margin-bottom: 55px;
  }
`;

export const LogoContainer = styled(Link)`
  height: 100%;
  width: 100px;
  padding: 25px;
  @media screen and (max-width: 800px) {
    padding: 0;
    width: 50px;
  }
  &:hover {
    transform: scale(1.15);
    transition: transform 0.6s cubic-bezier(0.25, 0.45, 0.45, 0.95);
  }
`;

export const OptionsContainer = styled.div`
  width: 50%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: flex-end;
  @media screen and (max-width: 800px) {
    margin: 10px;
    width: 80%;
  }
  @media screen and (max-width: 300px) {
    width: 100%;
    justify-content: space-around;
  }
`;

export const CartDropdownContainer = styled.div`
  position: absolute;
  top: 90px;
  right: ${global_padding_left_right}px;
  z-index: 5;
  @media screen and (max-width: 300px) {
    top: 120px;
    right: 16px;
  }
`;

export const OptionLink = styled(Link)`
  padding: 10px 15px;
  cursor: pointer;
  color: rgba(255, 255, 255, 1);
  &:hover {
    transform: scale(1.35);
    color: #ff6e7f;
    transition: transform 0.6s cubic-bezier(0.25, 0.45, 0.45, 0.95);
  }
`;
OptionLink.displayName = "OptionLink";
