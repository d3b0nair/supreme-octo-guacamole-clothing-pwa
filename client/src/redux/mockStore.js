import { createStore } from "redux";
import { persistReducer } from "redux-persist";
import { rootReducer, persistConfig } from "./root-reducer";
export const createMockStore = ({ mockState, mockReducers }) => {
  const combinedReducers = persistReducer(
    persistConfig,
    rootReducer(mockReducers)
  );
  return createStore(combinedReducers, mockState);
};
