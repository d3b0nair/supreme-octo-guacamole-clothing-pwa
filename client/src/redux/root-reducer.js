import { combineReducers } from "redux";
import { persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";

import userReducer from "./user/user.reducer";
import cartReducer from "./cart/cart.reducer";
import directoryReducer from "./directory/directory.reducer";
import shopReducer from "./shop/shop.reducer";
import checkoutReducer from "./checkout/checkout.reducer";
import stripeReducer from "./stripe/stripe.reducer";
import orderReducer from "./order/order.reducer"

const persistConfig = {
  key: "root",
  storage,
  whitelist: ["cart", "stripe"],
};
const rootReducer = combineReducers({
  user: userReducer,
  cart: cartReducer,
  checkout: checkoutReducer,
  directory: directoryReducer,
  shop: shopReducer,
  stripe: stripeReducer,
  selectedOrder: orderReducer
});
export default persistReducer(persistConfig, rootReducer);
