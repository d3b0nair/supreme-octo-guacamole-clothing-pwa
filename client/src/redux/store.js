import { createStore, applyMiddleware, compose } from "redux";
import { persistStore } from "redux-persist";
import logger from "redux-logger";
import createSagaMiddleware from "redux-saga";

import rootReducer from "./root-reducer";
import rootSaga from "./root-saga";

const sagaMiddleware = createSagaMiddleware();

const middlewares = [sagaMiddleware];
const composeEnhancers =
  (process.env.NODE_ENV !== "production" &&
    typeof window !== "undefined" &&
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) ||
  compose;
if (process.env.NODE_ENV === "development") {
  middlewares.push(logger);
}
export const store = createStore(
  rootReducer,
  composeEnhancers(applyMiddleware(...middlewares))
);

sagaMiddleware.run(rootSaga);

export const persistor = persistStore(store);
