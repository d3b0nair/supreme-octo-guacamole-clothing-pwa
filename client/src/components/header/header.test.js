import React from "react";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import {
  render,
  fireEvent,
  waitFor,
  screen,
  cleanup,
} from "@testing-library/react";
import { createMockStore } from "../../redux/mockStore";
import { Header } from "./header.component";
import { Provider } from "react-redux";

describe("Header component", () => {
  const store = createMockStore({}, {});

  const mockSignOutStart = jest.fn();
  const mockProps = (otherProps) => ({
    ...otherProps,
    currentUser: {
      uid: "123",
    },
    signOutStart: mockSignOutStart,
  });
  const history = createMemoryHistory();
  history.push("/signin");
  afterEach(cleanup);
  const component = (mockProps) => (
    <Router history={history}>
      <Provider store={store}>
        <Header {...mockProps} />
      </Provider>
    </Router>
  );

  it("should mount", () => {
    const { container } = render(component(mockProps({ hidden: true })));
    expect(container).toMatchSnapshot();
  });
  describe("if user loged in", () => {
    it("should call signOutStart method when link is clicked", async () => {
      render(component(mockProps()));
      const signOut = screen.getByTestId("signout-link");
      await waitFor(() => fireEvent.click(signOut));
      expect(mockSignOutStart).toHaveBeenCalled();
    });
  });

  describe("if hidden is false", () => {
    it("should render CartDropdown", () => {
      render(component(mockProps({ hidden: false })));
      const searchQuery = "Your cart is empty";
      const cartDropDown = screen.getByTestId("CartDropdown");
      expect(cartDropDown).toHaveTextContent(searchQuery);
    });
  });
});
