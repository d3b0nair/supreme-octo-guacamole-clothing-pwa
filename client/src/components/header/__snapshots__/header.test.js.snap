// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should mount 1`] = `
<div>
  <header
    class="sc-iqseJM bsgIPW"
  >
    <a
      class="sc-crHmcD duzzOY"
      href="/"
    >
      <svg
        alt="logo"
        class="logo"
      >
        logo.svg
      </svg>
    </a>
    <div
      class="sc-egiyK cOJwno"
    >
      <a
        class="sc-ksdxgE pFRJN"
        href="/shop"
      >
        SHOP
      </a>
      <a
        class="sc-ksdxgE pFRJN"
        href="/shop"
      >
        CONTACT
      </a>
      <div
        class="sc-ksdxgE pFRJN"
        data-testid="signout-link"
      >
        SIGN OUT
      </div>
      <div
        class="sc-bdvvtL jTehkV"
      >
        <svg
          class="sc-gsDKAQ tKSVz"
        >
          shopping-cart.svg
        </svg>
        <span
          class="sc-dkPtRN hSsJJR"
        >
          0
        </span>
      </div>
    </div>
  </header>
</div>
`;
